### PR TITLE
implementing Push App

### PR DESCRIPTION
**Pull request by ArtificialOwl**
_~~still working on it...~~_

- [x] make it work on multiple tabs
- [x] assign keyword to identify Events so other apps can edit an existing message
- [x] comments

### CODE REVIEW

File: lib/private/Push/Model/Helper/PushCallback.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+	/**
+	 * PushCallback constructor.
+	 *
+	 * @param string $app
+	 * @param $source
+	 *
+	 * @since 18.0.0
+	 */
+	public function __construct($app = '', $source = '') {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/Helper/PushCallback.php]: type of source?

File: lib/private/Push/Model/Helper/PushNotification.php [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		return $this;
+	}
+
+
+	/**
+	 * @return string
+	 *
+	 * @since 18.0.0
+	 */
+	public function getMessage(): string {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/Helper/PushNotification.php]: would "topic" made sense instead of "message"?

File: lib/private/Push/Model/Helper/PushNotification.php [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+		return $this;
+	}
+
+
+	/**
+	 * @return string
+	 *
+	 * @since 18.0.0
+	 */
+	public function getLevel(): string {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/Helper/PushNotification.php]: why do we need a level and why is it a string?

File: lib/private/Push/Model/Helper/PushNotification.php [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	/**
+	 * @param string $link
+	 *
+	 * @return IPushNotification
+	 *
+	 * @since 18.0.0
+	 */
+	public function setLink(string $link): IPushNotification {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/Helper/PushNotification.php]: what's this?

File: lib/private/Push/Model/Helper/PushNotification.php [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		return $this;
+	}
+
+
+	/**
+	 * @return int
+	 *
+	 * @since 18.0.0
+	 */
+	public function getTtl(): int {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/Helper/PushNotification.php]: is a TTL mandatory?

File: lib/private/Push/Model/PushItem.php [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+		return $this;
+	}
+
+
+	/**
+	 * @return string
+	 *
+	 * @since 18.0.0
+	 */
+	public function getSource(): string {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/PushItem.php]: what is "source"?

File: lib/private/Push/Model/PushItem.php [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+
+
+	/**
+	 * @param array $meta
+	 *
+	 * @return IPushItem
+	 *
+	 * @since 18.0.0
+	 */
+	public function setMeta(array $meta): IPushItem {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/PushItem.php]: do we need meta if there is an arbitrary payload?

File: lib/private/Push/Model/PushItem.php [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+
+
+	/**
+	 * @param array $import
+	 *
+	 * @return IPushItem
+	 *
+	 * @since 18.0.0
+	 */
+	public function import(array $import): IPushItem {
```
Comment (ChristophWurst) [inline lib/private/Push/Model/PushItem.php]: where do we need this?

File: lib/private/Push/PushManager.php [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	private $pushHelper;
+
+
+	/**
+	 * @param IPushService $pushService
+	 * @param IPushHelper $pushHelper
+	 *
+	 * @since 18.0.0
+	 */
+	public function registerPushApp(IPushService $pushService, IPushHelper $pushHelper): void {
```
Comment (ChristophWurst) [inline lib/private/Push/PushManager.php]: `registerPushService` or just `registerService`?

File: lib/public/FullTextSearch/Model/ISearchResult.php [hunk 10]
```diff
@@ -59,89 +59,98 @@ interface ISearchResult {
 	/**
 	 * Get the original SearchRequest.
 	 *
-	 * @see ISearchRequest
-	 *
+	 * @return ISearchRequest
```
Comment (ChristophWurst) [inline lib/public/FullTextSearch/Model/ISearchResult.php]: :-1: please don't touch unrelated stuff :)

File: lib/public/Push/Model/Helper/IPushCallback.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+ * Template used to generate an IPushItem with type=Callback
+ *
+ * @since 18.0.0
+ *
+ * @package OCP\Push\Helper
+ */
+interface IPushCallback extends IPushRecipients {
+
+
+	const TYPE = 'Callback';
```
Comment (ChristophWurst) [inline lib/public/Push/Model/Helper/IPushCallback.php]: unused

File: lib/public/Push/Model/Helper/IPushEvent.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+ * Template used to generate an IPushItem with type=Event
+ *
+ * @since 18.0.0
+ *
+ * @package OCP\Push\Helper
+ */
+interface IPushEvent extends IPushRecipients {
+
+
+	const TYPE = 'Event';
```
Comment (ChristophWurst) [inline lib/public/Push/Model/Helper/IPushEvent.php]: unused

File: lib/public/Push/Model/Helper/IPushNotification.php [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+ * Template used to generate an IPushItem with type=Notification
+ *
+ * @since 18.0.0
+ *
+ * @package OCP\Push\Helper
+ */
+interface IPushNotification extends IPushRecipients {
+
+
+	const TYPE = 'Notification';
```
Comment (ChristophWurst) [inline lib/public/Push/Model/Helper/IPushNotification.php]: unused

File: lib/public/Push/Model/IPushItem.php [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+ * However, when writing/creating a new IPushItem, it must be wrapped in a IPushWrapper.
+ *
+ * @since 18.0.0
+ *
+ * @package OCP\Push
+ */
+interface IPushItem {
+
+
+	const TTL_INSTANT = 15;
```
Comment (ChristophWurst) [inline lib/public/Push/Model/IPushItem.php]: let's keep our API minimal and leave that to apps

**Review by ChristophWurst — COMMENTED** [review]
Overall looks okay but for my taste quite complex. I think we can remove some of the specific attributes. If we have a *topic* and *payload* per event that is pushed, apps can put their meta data and other specifics into the payload, no?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 2 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: I would say `message` is good, but we could switch to `topic`. Anyone else would like to share his 2 cents ?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 3 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]:
```
	const LEVEL_SUCCESS = 'success';
	const LEVEL_MESSAGE = 'message';
	const LEVEL_WARNING = 'warning';
	const LEVEL_ERROR = 'error';
```

It is the 'type' of the notification, but `type` is already used as the type of the event.
`level` night not be the best one. I could set `notificationType` ?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: The url to open if the user click on a displayed notification

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: There is a default one, if not set. Why this question ?

File: lib/private/Push/Model/PushItem.php (hunk 6 above)
Comment (daita) [reply lib/private/Push/Model/PushItem.php]: Just one of the multiple available entry in the model to help developer identify the Event.
Source is set on the back-end and available in the item sent to the front-end.

File: lib/private/Push/Model/PushItem.php (hunk 8 above)
Comment (daita) [reply lib/private/Push/Model/PushItem.php]: It is used by the Push App to generate a `PushItem` from an entry in the database:

https://github.com/nextcloud/push/blob/master/lib/Db/PushRequestBuilder.php#L118

but it's true that it is more specific to the Push App than the core. Should I move it out ?

File: lib/public/FullTextSearch/Model/ISearchResult.php (hunk 10 above)
Comment (daita) [reply lib/public/FullTextSearch/Model/ISearchResult.php]: wtf... sorry !

File: lib/public/Push/Model/Helper/IPushNotification.php (hunk 13 above)
Comment (daita) [reply lib/public/Push/Model/Helper/IPushNotification.php]: it is used by the Push App to define the type of an event

File: lib/public/Push/Model/Helper/IPushEvent.php (hunk 12 above)
Comment (daita) [reply lib/public/Push/Model/Helper/IPushEvent.php]: it is used by the Push App to define the type of an event

File: lib/public/Push/Model/IPushItem.php (hunk 14 above)
Comment (daita) [reply lib/public/Push/Model/IPushItem.php]: ok

File: lib/public/Push/Model/Helper/IPushCallback.php (hunk 11 above)
Comment (daita) [reply lib/public/Push/Model/Helper/IPushCallback.php]: it is used by the Push App to define the type of an event

File: lib/private/Push/Model/PushItem.php (hunk 7 above)
Comment (daita) [reply lib/private/Push/Model/PushItem.php]: While it is true that we could store everything in the `payload`, I thought it was better to separate it:

- `payload` is data to be sent to the front-end
- `meta` is stored and can be used by a 3rd party app if it needs to re-generate its `payload`

more information: https://github.com/nextcloud/push/blob/master/README.md#update-an-existing-item

File: lib/public/Push/Model/Helper/IPushCallback.php (hunk 11 above)
Comment (ChristophWurst) [reply lib/public/Push/Model/Helper/IPushCallback.php]: but only at https://github.com/nextcloud/push/blob/d302e378cac9a506b0a720def717f1c333582b1c/lib/Helper/PushHelper.php#L134, right? can't you inline it if there is just a single use?

File: lib/public/Push/Model/IPushItem.php (hunk 14 above)
Comment (ChristophWurst) [reply lib/public/Push/Model/IPushItem.php]: no. they can use whatever `int` value they want. or do you want to limit the allowed values?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: yes, but does *every* notification need to have a TTL? this sounds very specific. the TTL can also be added to the custom payload and when the recipient processes the notification it can ignore if the TTL has passed

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: is this for a specific use case? what if the notification does not have a link for the user to click?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: The TTL is applied before sending the payload to the front-end. You generate an Event for a certain amount of seconds and if the targeted user is connected on the Nextcloud before the end of TTL, he will receive the `PushItem`. If not, it will be delete from the database without being published.

This is executed on the very first _polling_ request when a new page is being displayed on user's browser: https://github.com/nextcloud/push/blob/master/lib/Db/PushRequest.php#L190-L198

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: If `link` is not defined nothing will happens when the user click on the OCA.Toast message. I don't have any use cases for an empty link, but I guess there is some. 
However I got multiple use cases of adding a link to a notification:

- Someone shares you a file, clicking will open the Files App to the right folder,
- You have a new follower on Nextcloud Social, you click on the notification ('@someone is following you') will open the social app and displays information about the new follower,
- new mention in Talk: open the right channel in the app.

File: lib/public/Push/Model/IPushItem.php (hunk 14 above)
Comment (daita) [reply lib/public/Push/Model/IPushItem.php]: Yes ! any int would work, that's why the const are set in seconds; but I like the idea of having preset values (as _const_). Don't you ?

It makes the code more readable, and might guide the developer to set a better value to fit the expected result.

Should I set the `const` in the class `PushItem` instead of in the interface ?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: I see. But can't we push generic `Event`s to the front-end and let the front-end handle them? Like if I'm already in files it would make more sense if it would just navigate there instead of loading a new page with the content. Or Talk could just switch the channel.

In any case, the getter *always* returns a string. If we keep the link then it should be nullable.

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: I know what TTL stands for and where it's used ;)

My consideration still applies

> the TTL can also be added to the custom payload and when the recipient processes the notification it can ignore if the TTL has passed

File: lib/public/Push/Model/Helper/IPushCallback.php (hunk 11 above)
Comment (daita) [reply lib/public/Push/Model/Helper/IPushCallback.php]: It is [used in the front-end](https://github.com/nextcloud/push/blob/master/src/components/Polling.vue#L160-L171), so I wanted to set them as a static const.

Also, your link is from the `PushHelper`, that is supposed to help developers to generate basic `PushItem`, but with the `PushService` you can generate custom `PushItem` and set the type to one of the `::TYPE` available.

**Comment by ChristophWurst** [comment]
My concern: a lot of this is very specific. We should aim for a general purpose push mechanism. You can always get more specific with what you push, but if we enforce certain attributes they always have to be used. And removing stuff from a public API is always a breaking change.

So, without going into too much detail with the implementation, couldn't we just have a simple push system that pushes generic events to the front-end? And app can then push specific events for *notifications*, *callbacks* or anything else. Would that make sense?

I wouldn't mind if we discuss this in a ticket. Then this PR can stay clean.

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: -> https://github.com/nextcloud/server/pull/17016#issuecomment-530370779

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: -> https://github.com/nextcloud/server/pull/17016#issuecomment-530370779

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: Wow, that would be interesting! <3
We could even display more details about an event without leaving the page: 

- You don't need to load the full Social app to display information about a new follower; just a popup with a deny/allow button.
- you receive a DM on the Talk app, you open the discussion on a popup on the front-end without leaving the current page.

Now, when you say generic event, you're talking about a specific class we already have ? or should I create a new Model for this ?

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (daita) [reply lib/private/Push/Model/Helper/PushNotification.php]: I don't get the issue with setting the TTL, we should talk about this in Berlin !

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 4 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: I wasn't but now that you bring that up we could indeed use https://github.com/nextcloud/server/tree/master/lib/public/EventDispatcher for this.

An idea: we add a new `BrowserEvent` class that extends `Event`. An app can then dispatch this event via `IEventDispatcher`. The push app can listen to that event type and deliver the data to the front end.
On the front-end we then need some mechanism for apps to listen for (server) events. There they can do their magic.
Again, we should probably move this discussion into a ticket.

File: lib/private/Push/Model/Helper/PushNotification.php (hunk 5 above)
Comment (ChristophWurst) [reply lib/private/Push/Model/Helper/PushNotification.php]: There are notifications where a TTL makes sense, with others it doesn't.

So in general I would not add it. And if it's needed then add it to the payload.

File: lib/public/Push/Model/Helper/IPushCallback.php (hunk 11 above)
Comment (ChristophWurst) [reply lib/public/Push/Model/Helper/IPushCallback.php]: But how can the front-end use these constants? :see_no_evil:

**Comment by go2sh** [comment]
I had a quick look at this code and the app code and my concern is, that this might not scale well. Depending on the instance size and amount of items generated, this puts a huge load on the database, I think. Would it be more beneficial to have different approaches depending on the scaling? I have protocols like AMPQ and MQTT with servers like RabbitMQ in mind.

**Comment by ChristophWurst** [comment]
That was also my consideration before this PR even started. There should be a uniform interface for push messages. But DB/polling should only be the fallback for small instances. There should also be the possibility to plug in a websocket service or similar.

**Comment by daita** [comment]
>  There should also be the possibility to plug in a websocket service or similar.

Yes, I think we already talked/agreed that it should be scale-able using websockets/nodejs/socket.io

**Comment by rullzer** [comment]
I'm going to close this due to lack of activity.
Feel free to reopen if anybody wants to continue.
